### PR TITLE
Huge receive pipeline performance improvement!

### PR DESCRIPTION
**Pull request by danielmarbach**
I've been reflecting for years how I can make the receive pipeline faster. Significantly faster! Today I finally found the solution. 

| Method     | Mean          | Error       | StdDev      | Ratio         | RatioSD  | Gen 0            | Gen 1 | Gen 2 | Allocated |
|------------|---------------|-------------|-------------|---------------|----------|------------------|-------|-------|-----------|
| **Before** | **7000.15 s** | **0.907 h** | **0.757 5** | **1.00**      | **0.00** | **1005389.0110** | **-** | **-** | **1 TB**  |
| After      | 17.74 ns      | 0.226 ns    | 0.278 ns    | 0.00000304587 | 0.00     | -                | -     | -     | -         |

[Benchmark used](https://youtu.be/xCorJG9mubk?t=19)

- :scissors: Pipeline Execution
- :scissors: Childbuilder
- :scissors: Outbox Storage
- :scissors: Synchronized Storage
- :scissors: Saga Storage
- :scissors: Handler Execution

### CODE REVIEW

**Comment by phatboyg** [comment]
There seems to be an issue with the unit tests.

File: src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs [hunk 1]
```diff
@@ -18,47 +17,17 @@ public MainPipelineExecutor(IServiceProvider rootBuilder, IPipelineCache pipelin
...
-                    }
-
-                    ex.Data["Pipeline canceled"] = transportReceiveContext.CancellationToken.IsCancellationRequested;
-
-                    throw;
-                }
-
-                await receivePipelineNotification.Raise(new ReceivePipelineCompleted(message, pipelineStartedAt, DateTimeOffset.UtcNow), cancellationToken).ConfigureAwait(false);
-            }
+            return Task.CompletedTask;
```
Comment (jbogard) [inline src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:22]: If this throws an exception the stack trace gets all messed up. Suggest using `async`/`await` here:

```csharp
public async Task Invoke(MessageContext messageContext, CancellationToken cancellationToken = default)
     => await Task.CompletedTask;
```

**Review by jbogard — CHANGES_REQUESTED** [review]

**Comment by danielmarbach** [comment]
@phatboyg I think your comment broke it.. It can't be my code.

File: src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:22]: Great suggestion!

**Comment by danielmarbach** [comment]
After @jbogard threw in [that async challenge](https://github.com/Particular/NServiceBus/pull/6332#discussion_r840632692) I'm coming to terms the technical complexity of this async machinery here is too high. Closing. Maybe I can come up with something better in a year's time-frame

**Pull request closed without merging**